### PR TITLE
[docs][pytorch] Add examples for compiling with external weights.

### PR DESCRIPTION
**Pull request by vinayakdsci**
Progress on https://github.com/iree-org/iree/issues/18564.

Adds examples to the PyTorch guide, showing how to externalize module parameters, and load them at runtime, both through command line (`iree-run-module`) and through the iree-runtime Python API (using `ParameterIndex`).

### CODE REVIEW

File: docs/website/docs/guides/ml-frameworks/pytorch.md [hunk 1]
```diff
@@ -322,6 +322,132 @@ their values independently at runtime.
         self.value = new_value
     ```
 
+#### :octicons-code-16: Separating parameters during compilation
```
Comment (ScottTodd) [inline docs/website/docs/guides/ml-frameworks/pytorch.md]: nit: could use `octicons-file-symlink-file-16` to match the icon used for https://iree.dev/guides/parameters/ in the nav: https://github.com/iree-org/iree/blob/c86b621665ad8b898c0d0e987a45267f78eb9c3f/docs/website/docs/guides/parameters.md?plain=1#L2

Consistency and repetition across pages can make concepts easier to learn.

Comment (ScottTodd) [inline docs/website/docs/guides/ml-frameworks/pytorch.md]: nit: This isn't _just_ about compilation, it's also about runtime, since different parameters can be fed to the same loaded program at runtime. Could change the section header to something more like "Using external parameters"

File: docs/website/docs/guides/ml-frameworks/pytorch.md [hunk 2]
```diff
@@ -322,6 +322,132 @@ their values independently at runtime.
...
+
+In practical scenarios, it is usually unfeasible to embed model parameters
+in the IR.
+
+`aot.externalize_module_parameters()` allows for the separation of
+the parameters from the IR, but still encodes a symbolic relationship between
+between parameters and the IR, so they can be loaded at runtime.
```
Comment (ScottTodd) [inline docs/website/docs/guides/ml-frameworks/pytorch.md]: This first sentence is a bit awkward ("unfeasible") and focuses on internal implementation details ("IR" intermediate representation is mostly internal to the compiler).

Borrowing some language from https://iree.dev/guides/parameters/, I'd rephrase to focus on what benefits external parameters provide and maybe use terms like "program"/"code"/"compute graph" instead of "IR". Acronyms can also be overwhelming for users, especially when the acronyms come from different domains.

This should also link to `../parameters.md` for more information somewhere near the top.

```suggestion
Model parameters can be stored in standalone files that can be efficiently
stored and loaded separately from model compute graphs. See the
[Parameters guide](../parameters.md) for more general information about
parameters in IREE.

When using iree-turbine, the `aot.externalize_module_parameters()` function
separates parameters from program modules and encodes a symbolic relationship
between them so they can be loaded at runtime.
```

File: docs/website/docs/guides/ml-frameworks/pytorch.md [hunk 3]
```diff
@@ -322,6 +322,132 @@ their values independently at runtime.
         self.value = new_value
     ```
 
+#### :octicons-code-16: Separating parameters during compilation
+
+In practical scenarios, it is usually unfeasible to embed model parameters
```
Comment (zjgarvey) [inline docs/website/docs/guides/ml-frameworks/pytorch.md]: nit: infeasible is more common. 

nit: instead of "In practical scenarios", maybe put "For large models", since this is the practical scenario where this applies.

File: docs/website/docs/guides/ml-frameworks/pytorch.md [hunk 4]
```diff
@@ -322,6 +322,132 @@ their values independently at runtime.
...
+# Create a params dictionary. Note that the keys here match
+# LinearModule's attributes. We will use the saved safetensor
+# file for use from the command line.
+from safetensors.torch import save_file
+
+wt = torch.randn(4,3)
+bias = torch.randn(3)
+params = { "weight": wt, "bias": bias }
+save_file(params, "params.safetensors")
+```
```
Comment (zjgarvey) [inline docs/website/docs/guides/ml-frameworks/pytorch.md]: I'm finding this section misleading from the example above. The `linear_module` already has specific weight and bias parameters. Why are we storing new random ones?

File: docs/website/docs/guides/ml-frameworks/pytorch.md [hunk 5]
```diff
@@ -322,6 +322,132 @@ their values independently at runtime.
...
+    def __init__(self, in_features, out_features):
+        super().__init__()
+        self.weight = torch.nn.Parameter(torch.randn(in_features, out_features))
+        self.bias = torch.nn.Parameter(torch.randn(out_features))
+
+    def forward(self, input):
+        return (input @ self.weight) + self.bias
+
+# Construct LinearModule.
+linear_module = LinearModule(4,3)
```
Comment (ScottTodd) [inline docs/website/docs/guides/ml-frameworks/pytorch.md]: Drop obvious comment to keep the code sample smaller

```suggestion
linear_module = LinearModule(4,3)
```

File: docs/website/docs/guides/ml-frameworks/pytorch.md [hunk 6]
```diff
@@ -322,6 +322,132 @@ their values independently at runtime.
...
+
+```python
+import numpy as np
+import shark_turbine.aot as aot
+
+# Externalize the model parameters. This removes weight
+# tensors from the IR module, allowing them to be loaded
+# at runtime. Symbolic references to these parameters are
+# still retained in the IR.
+aot.externalize_module_parameters(linear_module)
```
Comment (zjgarvey) [inline docs/website/docs/guides/ml-frameworks/pytorch.md:367]: Where does this externalize the weights to? Does the file have to be called `params.safetensors` in the PWD or something?

File: docs/website/docs/guides/ml-frameworks/pytorch.md [hunk 7]
```diff
@@ -322,6 +322,132 @@ their values independently at runtime.
...
+
+In practical scenarios, it is usually unfeasible to embed model parameters
+in the IR.
+
+`aot.externalize_module_parameters()` allows for the separation of
+the parameters from the IR, but still encodes a symbolic relationship between
+between parameters and the IR, so they can be loaded at runtime.
+
+We define a `torch.nn.Module` that will be compiled, and save
+random tensors to be loaded as parameters in a safetensors file.
```
Comment (ScottTodd) [inline docs/website/docs/guides/ml-frameworks/pytorch.md]: I would link either https://huggingface.co/docs/safetensors/ or https://github.com/huggingface/safetensors here when introducing "safetensors" for context.

Like this: `[safetensors](https://huggingface.co/docs/safetensors/)`

File: docs/website/docs/guides/ml-frameworks/pytorch.md [hunk 8]
```diff
@@ -322,6 +322,132 @@ their values independently at runtime.
...
+import shark_turbine.aot as aot
+
+# Externalize the model parameters. This removes weight
+# tensors from the IR module, allowing them to be loaded
+# at runtime. Symbolic references to these parameters are
+# still retained in the IR.
+aot.externalize_module_parameters(linear_module)
+
+ext_arg = torch.randn(4)
+exported_module = aot.export(linear_module, ext_arg)
```
Comment (zjgarvey) [inline docs/website/docs/guides/ml-frameworks/pytorch.md]: What is ext_arg? Is this an example input?

File: docs/website/docs/guides/ml-frameworks/pytorch.md [hunk 9]
```diff
@@ -322,6 +322,132 @@ their values independently at runtime.
...
+
+# Compile the exported module, to generate the binary.
+binary = exported_module.compile(save_to=None)
+
+# Save the input as an npy tensor, for later reuse.
+ext_arg_2 = ext_arg.numpy()
+np.save("input.npy", ext_arg_2)
```
Comment (zjgarvey) [inline docs/website/docs/guides/ml-frameworks/pytorch.md]: Ah, `ext_arg` is an input! I'd put a comment before you define `ext_arg` above.

File: docs/website/docs/guides/ml-frameworks/pytorch.md [hunk 10]
```diff
@@ -322,6 +322,132 @@ their values independently at runtime.
...
+in to IREE through the command line.
+
+```bash
+iree-run-module --module=compiled_module.vmfb --parameters=model=params.safetensors \
+                --input=@input.npy
+```
+
+Note here that the `--parameters` flag has `model=` following it immediately.
+This simply specifies the scope of the parameters, and is reflected in the
+compiled IR.
```
Comment (zjgarvey) [inline docs/website/docs/guides/ml-frameworks/pytorch.md]: I'm not sure I understand this. In what situations are parameters *not* model wide?

File: docs/website/docs/guides/ml-frameworks/pytorch.md [hunk 11]
```diff
@@ -322,6 +322,132 @@ their values independently at runtime.
...
+in the IR.
+
+`aot.externalize_module_parameters()` allows for the separation of
+the parameters from the IR, but still encodes a symbolic relationship between
+between parameters and the IR, so they can be loaded at runtime.
+
+We define a `torch.nn.Module` that will be compiled, and save
+random tensors to be loaded as parameters in a safetensors file.
+
+```python
```
Comment (ScottTodd) [inline docs/website/docs/guides/ml-frameworks/pytorch.md:339]: To keep the overall page organized, I would either put this entire example into a single code block:
```python
# Define a torch.nn.Module with some parameters.
import torch
class LinearModule(torch.nn.Module):
    ...

# Save the parameters to a .safetensors file.
...
save_file(params, "params.safetensors")

# Externalize the model parameters.
aot.externalize_module_parameters(linear_module)
...
```

or break out each step into an ordered or unordered list:

````
1. We define a `torch.nn.Module` that will be compiled, and save
random tensors to be loaded as parameters in a safetensors file.

    ```python
    import torch
    ...
````

Here's how the page would look with a list:
image

File: docs/website/docs/guides/ml-frameworks/pytorch.md [hunk 12]
```diff
@@ -322,6 +322,132 @@ their values independently at runtime.
...
+linear_module = LinearModule(4,3)
+```
+
+Safetensors can be created directly from a parameters dictionary created from
+randomly generated torch tensors.
+
+```python
+# Create a params dictionary. Note that the keys here match
+# LinearModule's attributes. We will use the saved safetensor
+# file for use from the command line.
```
Comment (ScottTodd) [inline docs/website/docs/guides/ml-frameworks/pytorch.md]: You can wrap these lines at closer to 80 characters if you want. Still quite a bit of unused whitespace:

image

**Review by zjgarvey — COMMENTED** [review]
Nice, I'm getting a sense for how this external parameter thing is working. Some comments based on a fairly naive read-through.

File: docs/website/docs/guides/ml-frameworks/pytorch.md (hunk 8 above)
Comment (ScottTodd) [inline docs/website/docs/guides/ml-frameworks/pytorch.md]: What's this `ext_arg`? In sample code you can be more verbose if it makes it easier to follow.

File: docs/website/docs/guides/ml-frameworks/pytorch.md (hunk 9 above)
Comment (zjgarvey) [reply docs/website/docs/guides/ml-frameworks/pytorch.md]: maybe rename ext_arg entirely?

File: docs/website/docs/guides/ml-frameworks/pytorch.md (hunk 6 above)
Comment (zjgarvey) [reply docs/website/docs/guides/ml-frameworks/pytorch.md:367]: Reading further on, it looks like you specify the params file when invoking iree-run-module. I suppose the symbolic reference is to the name of the parameter in the safetensors dictionary you saved earlier?

File: docs/website/docs/guides/ml-frameworks/pytorch.md [hunk 13]
```diff
@@ -322,6 +322,132 @@ their values independently at runtime.
...
+binary = exported_module.compile(save_to="compiled_module.vmfb")
+```
+
+The stored safetensors file, the input tensor, and the VMFB can now be passed
+in to IREE through the command line.
+
+```bash
+iree-run-module --module=compiled_module.vmfb --parameters=model=params.safetensors \
+                --input=@input.npy
+```
```
Comment (ScottTodd) [inline docs/website/docs/guides/ml-frameworks/pytorch.md]: The sample code you have here takes a branch, so you might want to consider using content tabs: https://squidfunk.github.io/mkdocs-material/reference/content-tabs/

* Compiler (always Python, since this is PyTorch)
* Runtime
  * Python API
  * C API
  * Native tools

Steps for each runtime choice would go in a separate tab:

image

File: docs/website/docs/guides/ml-frameworks/pytorch.md (hunk 4 above)
Comment (vinayakdsci) [reply docs/website/docs/guides/ml-frameworks/pytorch.md]: Just as an example! We might not always have the class that actually represents the model, but just some representation of the graph. This is a demonstration of how, in such a case, one can load the weights separately.

File: docs/website/docs/guides/ml-frameworks/pytorch.md (hunk 10 above)
Comment (vinayakdsci) [reply docs/website/docs/guides/ml-frameworks/pytorch.md]: model is a _scope_ here. It could actually be anything. If is even possible that the biases could be under a scope called "b", and weights under a scope called "a". In that case, they would be in different indices, and `model=` would be then `b=` and `a=` respectively.

File: docs/website/docs/guides/ml-frameworks/pytorch.md (hunk 6 above)
Comment (vinayakdsci) [reply docs/website/docs/guides/ml-frameworks/pytorch.md:367]: No! It can be any path you specify. Here it is this file in PWD.

File: docs/website/docs/guides/ml-frameworks/pytorch.md (hunk 3 above)
Comment (vinayakdsci) [reply docs/website/docs/guides/ml-frameworks/pytorch.md]: Ok, will do.

**Review by ScottTodd — COMMENTED** [review]
Sending my initial comments since a few already overlap with Zach's :)

File: docs/website/docs/guides/ml-frameworks/pytorch.md (hunk 9 above)
Comment (vinayakdsci) [reply docs/website/docs/guides/ml-frameworks/pytorch.md]: Ok, will do.

File: docs/website/docs/guides/ml-frameworks/pytorch.md (hunk 6 above)
Comment (vinayakdsci) [reply docs/website/docs/guides/ml-frameworks/pytorch.md:367]: > Reading further on, it looks like you specify the params file when invoking iree-run-module. I suppose the symbolic reference is to the name of the parameter in the safetensors dictionary you saved earlier?

Yes, and under the scope named "model".

**Comment by stellaraccident** [comment]
Thanks for doing this. Once landed, can someone make sure to add a test to iree-turbine which does this exact thing?

**Comment by vinayakdsci** [comment]
@stellaraccident Sure! I'd be happy to do that.

**Review by ScottTodd — APPROVED** [review]
Great! Thanks!

File: docs/website/docs/guides/ml-frameworks/pytorch.md (hunk 4 above)
Comment (vinayakdsci) [reply docs/website/docs/guides/ml-frameworks/pytorch.md]: @zjgarvey I changed the examples to use the modules parameters.

**Review by zjgarvey — APPROVED** [review]
Nice, this looks good to me. Thanks for the changes!
